### PR TITLE
[1238] Dttp retrieve contact

### PR DESCRIPTION
**Pull request by stellenberger**
### Context

- [Trello link](https://trello.com/c/S3wwHxHE/1238-m-retrieve-contact)
- Retrieve filtered contacts to cross reference using trainee dttp_id (contactid on Dttp)

### Changes proposed in this pull request

- Created a RetrieveContact class inside of Dttp module. 
- Constant `CONTACT_ATTRIBUTE_FILTERS` holds filter values for Dttp
- Call method retrieves the filtered attributes of the contact requested. Needs a trainee with valid `dttp_id`

### Guidance to review

1. Enter your rails console `$ rails c`
2. Create a new trainee with a valid dttp_id, depending on environment `trainee = Trainee.new(dttp_id: [ENTER_VALID_ID], provider: Provider.last, training_route: 1)`
3. Make the retrieve call through Dttp Module `Dttp::Contact.call(trainee: trainee)`

NOTE: To get a valid dttp_id, you may need to go into postman to get it. Go to the `contacts` endpoint and get an example `contactid`. Use `dev` environment

### CODE REVIEW

**Comment by avinhurry** [comment]
[[M] Retrieve contact](https://trello.com/c/S3wwHxHE/1238-m-retrieve-contact)

File: spec/services/dttp/retrieve_contact_spec.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        allow(Client).to receive(:get).with(path).and_return(dttp_response)
+      end
+
+      context "filtered params" do
+        let(:parsed_response) do
+          {
+            "contactid" => "XXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
+            "firstname" => "John",
+            "lastname" => "Smith",
+          }
```
Comment (defong) [inline spec/services/dttp/retrieve_contact_spec.rb:24]: missing `emailaddress1`
```suggestion
           "emailaddress1" => ""email@example.com,
          }
```

**Review by defong — CHANGES_REQUESTED** [review]

File: app/services/dttp/retrieve_contact.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    class HttpError < StandardError; end
+
+    attr_reader :trainee
+
+    CONTACT_ATTRIBUTE_FILTERS = %w[
+      emailaddress1
+      firstname
+      lastname
+      contactid
```
Comment (defong) [inline app/services/dttp/retrieve_contact.rb:15]:
```suggestion
```


https://github.com/DFE-Digital/register-trainee-teacher-data/blob/56726004e56996f94c6254e5902dc22553458dfd/app/services/dttp/register_for_trn.rb#L44-L46

https://github.com/DFE-Digital/register-trainee-teacher-data/blob/da0c9760ee1e0c2d410028819d937f0d0e66fb82/app/models/trainee.rb#L141-L143


`contactid` is the `trainee.dttp_id`

File: app/services/dttp/retrieve_contact.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      @trainee = trainee
+    end
+
+    def call
+      response = Client.get("/contacts(#{trainee.dttp_id})?$select=#{filters}")
+      if response.code != 200
+        raise HttpError, "status: #{response.code}, body: #{response.body}, headers: #{response.headers}"
+      end
+
+      JSON(response.body)
```
Comment (gpeng) [inline app/services/dttp/retrieve_contact.rb:27]: Do you think we should create a class to wrap these to abstract the DTTP JSON schema away from our app? Maybe `Dttp::Contact` that takes the JSON and then has properly named accessors.

Comment (stellenberger) [reply app/services/dttp/retrieve_contact.rb:27]: Sure, do you mean to switch the DTTP accessors (emailaddress1, firstname etc) to Trainee column names (email_address, first_name etc?)

Comment (gpeng) [reply app/services/dttp/retrieve_contact.rb:27]: Yes, I think we could try that. Just something that will abstract away all of the inconsistencies in the JSON naming would be good.

File: app/services/dttp/retrieve_contact.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+module Dttp
+  class RetrieveContact
```
Comment (gpeng) [inline app/services/dttp/retrieve_contact.rb:4]: This should be consistently named with the other two that @JonF1  and @avinhurry are doing. I'd probably go with `Dttp::Contact::Fetch`, `Dttp::PlacementAssignment::Fetch`, `Dttp::Account::Fetch` and then each of those returns a `Dttp::Contact`, `Dttp::PlacementAssignment`, `Dttp::Account`.

**Review by gpeng — APPROVED** [review]

**Review by defong — APPROVED** [review]

**Pull request merged**